File: balanced_bt.java
//TC:O(n)
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        if(root==null)
        return true;
        return(helper(root)!=-1);
    }
    private int helper(TreeNode root)
    {
        if(root==null)
        return 0;
        int left=helper(root.left);
        int right=helper(root.right);
        if(Math.abs(left-right)>1)
        return -1;
        if(left==-1||right==-1)
        return -1;
        return Math.max(left,right)+1;
    }
}

File: Balanced_BT.py
# Time complexity : O(n)
# Space complexity : O(n)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


class Solution:
    
    def check(self, root):
            
            if not root:
                return [True, 0]
            
            
            left = self.check(root.left)
            right = self.check(root.right)
            
            check_balanced = left[0] and right[0] and (abs(left[1] - right[1]) <= 1)
            
            
            return [check_balanced, 1 + max(left[1], right[1])]
        
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        
        return self.check(root)[0]
